src/types/binary/binaryType.ts

```diff
@@ -0,0 +1,303 @@
+   * @returns unsigned number with changed value
+   */
+  protected clearBitOnNumber(bitOffset: number): number {
+    this.throwErrorIfBitOffsetNotInRange(bitOffset)
+    return convertToUnsignedNumber(this.value & ~(1 << bitOffset))
+  }
+
+  /**
+   * sets the bit to 1 when it was 0 or to 0 if it was 1 before
+   * @param bitOffset
+   * @returns unsigned number with changed value
+   */
+  protected toggleBitOnNumber(bitOffset: number): number {
+    this.throwErrorIfBitOffsetNotInRange(bitOffset)
+    return this.isBitSet(bitOffset)
+      ? this.clearBitOnNumber(bitOffset)
+      : this.setBitOnNumber(bitOffset)
+  }
+
+  /**
+   * throws system exception if offset is not smaller than count of bits for type
+   * @param bitOffset 0-indexed offset
+   */
+  private throwErrorIfBitOffsetNotInRange(bitOffset: number): void {
+    if (bitOffset >= this.numberOfBitsForType || bitOffset < 0) {
+      throw new Error('bit offset (tried to access) is not within type range')
+    }
+  }
+}
+
+/**
+ * Represents a byte in range (0x00 - 0xFF).
+ */
+export class Byte extends BinaryType implements IByte {
+  public static MIN_VALUE: number = 0
+  public static MAX_VALUE: number = 0xff //decimal: 255
+  public static MIN_SIGNED_VALUE: number = -128
+  public static MAX_SIGNED_VALUE: number = 127
+  public static readonly ZERO_BYTE: Byte = Byte.fromUnsignedInteger(0)
+
+  readonly numberOfBitsForType: number = 8
+  readonly maxValueForType: number = 0xff
+
+  /**
+   * The unsigned integer representation of the byte as a number (IEEE double precision floating point).
+   */
+  public readonly value: number
+
+  private constructor(value: number) {
+    super()
+    checkRange('Byte', value, Byte.MIN_VALUE, Byte.MAX_VALUE)
+    this.value = value
+  }
+
+  /**
+   * Creates a new byte from an unsigned integer.
+   *
+   * @param value the unsigned integer value
+   * @returns the byte representation
+   */
+  public static fromUnsignedInteger(value: number): Byte {
+    checkRange('8-bit unsigned integer', value, Byte.MIN_VALUE, Byte.MAX_VALUE)
+    return new Byte(value)
+  }
+
+  /**
+   * Creates a new byte from a signed integer.
+   *
+   * @param value the signed integer value
+   * @returns the byte representation
+   */
+  public static fromSignedInteger(value: number): Byte {
+    checkRange(
+      '8-bit signed integer',
+      value,
+      Byte.MIN_SIGNED_VALUE,
+      Byte.MAX_SIGNED_VALUE
+    )
+    if (value < Byte.MIN_VALUE) {
+      return new Byte(Byte.MAX_VALUE + value + 1)
+    }
+    return new Byte(value)
+  }
+  /**
+   * Adds the specified number to the byte and returns the result as a new byte. In case the
+   * result exceeds the `Byte.MAX_VALUE` then it will overflow.
+   *
+   * @param value the number to be added to the byte
+   * @returns the new byte with the value added
+   */
+  public add(value: Byte | number): Byte {
+    return new Byte(
+      limitValuesToBitCount(this.addToNumber(value), this.numberOfBitsForType)
```

Comment:
Why do we not use the abstract max value and make an & operation with it? To me this seems to be a bit of an overhead. But I guess this is more personal preference.

Edit: I just saw that you do some additional checks like safeInteger etc. I guess then its fine.

Reply:
you are right, since with our 3 binary types we are always within safe integer range for add operation

Reply:
keep the unused methods `limitValuesToBitCount, getMaxValueDisplayableWithBits` perhaps can be used for something else in future